sdk/python/feast/infra/online_stores/datastore.py

```diff
@@ -185,16 +187,30 @@ def _write_minibatch(
         progress: Optional[Callable[[int], Any]],
     ):
         entities = []
-        for entity_key, features, timestamp, created_ts in data:
+        all_datastore_keys = []
+        for entity_key, _, _, _ in data:
             document_id = compute_entity_id(entity_key)
-
             key = client.key(
                 "Project", project, "Table", table.name, "Row", document_id,
             )
+            all_datastore_keys.append(key)
+        entities = client.get_multi(all_datastore_keys)
+        for entity_key, features, timestamp, created_ts in data:
+            document_id = compute_entity_id(entity_key)
 
-            entity = datastore.Entity(
-                key=key, exclude_from_indexes=("created_ts", "event_ts", "values")
+            key = client.key(
+                "Project", project, "Table", table.name, "Row", document_id,
             )
+            existing_entity: datastore.Entity = client.get(key)
+            new_ts = utils.make_tzaware(timestamp)
+            if existing_entity and existing_entity["event_ts"] >= new_ts:
+                continue
```

Comment:
Not entirely sure this would improve speed as we are still traversing through the same number, but we could remove L190-197 and merge the two traversals like this:
```suggestion
            if existing_entity and existing_entity["event_ts"] >= new_ts:
                key = client.key(
                    "Project", project, "Table", table.name, "Row", document_id,
                )
                entities.append(client.get(key))
                continue
```